pkg/edgeview/vendor/github.com/lf-edge/eve/pkg/pillar/types/assignableadapters.go

```diff
@@ -29,6 +29,100 @@ type AssignableAdapters struct {
 	IoBundleList []IoBundle
 }
 
+type ioBundleErrorBase struct {
+	ErrStr  string
+	TypeStr string
+}
+
+func (i ioBundleErrorBase) Error() string {
+	return i.ErrStr
+}
+
+// IOBundleError is an error stored in IoBundles that can be marshalled
+type IOBundleError struct {
+	Errors      []ioBundleErrorBase
+	TimeOfError time.Time
+}
+
+// ErrorTime returns the time of the last error added
+func (iobe *IOBundleError) ErrorTime() time.Time {
+	return iobe.TimeOfError
+}
+
+func (iobe *IOBundleError) String() string {
+	if len(iobe.Errors) == 0 {
+		return ""
+	}
+	errorStrings := make([]string, 0, len(iobe.Errors))
+	for _, err := range iobe.Errors {
+		errorStrings = append(errorStrings, err.Error())
+	}
+	return strings.Join(errorStrings, "; ")
+}
+
+// Append converts an error to ioBundleErrorBase and adds it
+func (iobe *IOBundleError) Append(err error) {
+	if iobe.Errors == nil {
+		iobe.Errors = make([]ioBundleErrorBase, 0, 1)
+	}
+
+	typeStr := reflect.TypeOf(err).String()
+	baseErr := ioBundleErrorBase{
+		ErrStr:  err.Error(),
+		TypeStr: typeStr,
+	}
+
+	iobe.Errors = append(iobe.Errors, baseErr)
+
+	iobe.TimeOfError = time.Now()
+}
+
+// Empty returns true if no error has been added
+func (iobe *IOBundleError) Empty() bool {
+	if iobe.Errors == nil || len(iobe.Errors) == 0 {
+		return true
+	}
+
+	return false
+}
+
+// HasErrorByType returns true if error of the same type is found
+func (iobe *IOBundleError) HasErrorByType(e error) bool {
```

Comment:
This is already in master: https://github.com/lf-edge/eve/blob/master/pkg/pillar/types/assignableadapters.go#L90

I don't understand why GH is showing this; yesterday @rucoder had a similar issue ...

Reply:
But this is a copy of that file vendored under edgeview package.
Or do you mean that we miss some recent change here?

Reply:
ah sorry, my mistake; then everything is fine.

Reply:
Can't edgeview just use `replace` in go.mode?

Reply:
@deitch can write an essay on why `replace` is evil :)

Reply:
> Can't edgeview just use `replace` in go.mod?

Do we use the old version there?... So the PR should be fixed?

I remember we had a reason to put our code under vendor (though I need to remember it exactly), but it still looks awful and has to be at least automated from the perspective of updating/checking in PRs...

Reply:
> write an essay

I can, you won't want to read it, though. :-)

Basically don't do it, except for local circumstances, like "I am developing something and I need to use my local copy of this other dependency, just to see if it all works together, but I will not commit it that way."

I don't understand what problem edgeview is trying to solve that it needs `replace`? Or is it the issue that `pkg/edgeview` and `pkg/pillar` each build independently, but there is common code to both, so you want to reuse one from the other?

Reply:
Just to not having to create these PRs and make it automatic - is there another way?

Reply:
Libraries, like [eve-libs](http://github.com/lf-edge/eve-libs/), or some sort of tooling that auto-generates, but those are painful. Pretty soon you are building bazel.

Reply:
I'm still confused about what we are doing with this PR. Should it be fixed?

Reply:
From my side it is fine then.

Reply:
> > write an essay
> 
> I can, you won't want to read it, though. :-)

Meanwhile Eden's go.mod : 
![](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExMGV2Z3h3bXlqdTQwdjN1N25pdm52dmRnMmJleTYzbGo5OGdxaGd4MCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/COYGe9rZvfiaQ/giphy.webp)